common/locks/priority_semaphore_impl.go

```diff
@@ -0,0 +1,136 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package locks
+
+import (
+	"context"
+	"sync"
+)
+
+type Priority int
+
+const (
+	PriorityHigh Priority = iota
+	PriorityLow
+)
+
+type PrioritySemaphoreImpl struct {
+	capacity    int
+	highWaiting int         // Current number of high priority requests waiting.
+	highCount   int         // Current count of high-priority holds
+	lowCount    int         // Current count of low-priority holds
+	lock        sync.Locker // Mutex to protect conditional variable and changes to the counts
+	highCV      *sync.Cond  // Condition variable for high-priority tasks
+	lowCV       *sync.Cond  // Condition variable for low-priority tasks
```

Comment:
these don't have to be pointers, you can just do
```suggestion
	lock        sync.Mutex // Mutex to protect conditional variable and changes to the counts
	highCV      sync.Cond  // Condition variable for high-priority tasks
	lowCV       sync.Cond  // Condition variable for low-priority tasks
```
and then set `sem.highCV.L = &sem.lock` in the constructor

Reply:
Nice! Made the changes.